lib/complate.js

```diff
@@ -0,0 +1,44 @@
+let { VirtualBundle } = require("faucet-pipeline-js/lib/bundle/virtual");
+
+// a plugin loads the highlighter for the language
+let loadLanguages = require("prismjs/components/");
+loadLanguages(["jsx"]);
+
+let rendering = `
+import Renderer, { Fragment as _F, createElement as _h } from "complate-stream";
+import BufferedStream from "complate-stream/src/buffered-stream";
+
+let renderer = new Renderer();
+
+function render(macro, fragment) {
+	let view = () => _h(_F, null, macro);
+	let stream = new BufferedStream();
+	renderer.renderView(view, null, stream, { fragment }, () => {
+		let html = stream.read();
+	RENDERED = html;
+	});
+}`;
+
+let bundle = new VirtualBundle("./", {
```

Comment:
So the reference directory is set to whatever the current working directory is? I've become increasingly wary of such implicit state, as it's hard to test, debug and, uhm, reason about 👨‍💻 .

I'm still insufficiently familiar with how everything else here fits together, but would recommend being explicit:

```javascript
let DEFAULT_CONFIG = { … };
let DEFAULT_OPTIONS = { browsers: {} };
let BUNDLES = new Map(); // global state required for caching
…
module.exports = function(snippet, referenceDir) {
    let bundle = BUNDLES.get(referenceDir);
    if(!bundle) {
        bundle = new VirtualBundle(referenceDir, DEFAULT_CONFIG, DEFAULT_OPTIONS);
        BUNDLES.set(referenceDir, bundle);
    }
    …
};
```

(not sure whether `referenceDir` makes for a suitable key here - perhaps we do just want a single bundle across all snippets; can't think it through right now)

Reply:
Agreed.